src/components/Form.js

```diff
@@ -0,0 +1,216 @@
+        if (this.props.formState.isSubmitting) {
+            return;
+        }
+
+        // Touches all form inputs so we can validate the entire form
+        _.each(this.inputRefs, (inputRef, inputID) => (
+            this.touchedInputs[inputID] = true
+        ));
+
+        // Validate form and return early if any errors are found
+        const values = this.getValues();
+        if (!_.isEmpty(this.validate(values))) {
+            return;
+        }
+
+        // Set loading state and call submit handler
+        FormActions.setIsSubmitting(this.props.formID, true);
+        this.props.onSubmit(values);
+    }
+
+    /**
+     * @param {Object} values - An object containing the value of each inputID, e.g. {inputID1: value1, inputID2: value2}
+     * @returns {Object} - An object containing the errors for each inputID, e.g. {inputID1: error1, inputID2: error2}
+     */
+    validate(values) {
+        FormActions.setServerErrorMessage(this.props.formID, '');
+        const validationErrors = this.props.validate(values);
+
+        if (!_.isObject(validationErrors)) {
+            throw new Error('Validate callback must return an empty object or an object with shape {inputID: error}');
+        }
+
+        const errors = _.pick(validationErrors, (inputValue, inputID) => (
+            Boolean(this.touchedInputs[inputID])
+        ));
+        this.setState({errors});
+        return errors;
+    }
+
+    childrenWrapperWithProps(children) {
+        return React.Children.map(children, (child) => {
+            // Just render the child if it is not a valid React element, e.g. text within a <Text> component
+            if (!React.isValidElement(child)) {
+                return child;
+            }
+
+            // Depth first traversal of the render tree as the input element is likely to be the last node
+            if (child.props.children) {
+                return React.cloneElement(child, {
+                    children: this.childrenWrapperWithProps(child.props.children),
+                });
+            }
+
+            // We check if the child has the isFormInput prop.
+            // We don't want to pass form props to non form components, e.g. View, Text, etc
+            if (!child.props.isFormInput) {
+                return child;
+            }
+
+            // We clone the child passing down all form props
+            const inputID = child.props.inputID;
+            return React.cloneElement(child, {
+                ref: node => this.inputRefs[inputID] = node,
+                defaultValue: this.props.draftValues[inputID] || child.props.defaultValue,
+                errorText: this.state.errors[inputID] || '',
+                onBlur: (event) => {
+                    this.setTouchedInput(inputID);
+                    this.validate(this.getValues());
+                    if (child.props.onBlur) {
+                        child.props.onBlur(event);
+                    }
```

Comment:
NAB. maybe there is some case to hook into the `onBlur` eventually, but we can probably add it later.